compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala

```diff
@@ -180,9 +176,189 @@ object ErrorReporting {
+
+  private def defaultAmbiguousImplicitMsg(ambi: AmbiguousImplicits) = {
+    formatMsg(s"ambiguous implicit arguments: ${ambi.explanation}${location("of")}")(
+      s"ambiguous implicit arguments of type ${pt.show} found${location("for")}"
+    )
+  }
+
+  private def defaultImplicitNotFoundMessage = {
+    em"no implicit argument of type $pt was found${location("for")}"
+  }
+
+  /** Construct a custom error message given an ambiguous implicit
+   *  candidate `alt` and a user defined message `raw`.
+   */
+  private def userDefinedAmbiguousImplicitMsg(alt: SearchSuccess, raw: String) = {
+    val params = alt.ref.underlying match {
+      case p: PolyType => p.paramNames.map(_.toString)
+      case _           => Nil
+    }
+    def resolveTypes(targs: List[tpd.Tree])(using Context) =
+      targs.map(a => Inferencing.fullyDefinedType(a.tpe, "type argument", a.span))
+
+    // We can extract type arguments from:
+    //   - a function call:
+    //     @implicitAmbiguous("msg A=${A}")
+    //     implicit def f[A](): String = ...
+    //     implicitly[String] // found: f[Any]()
+    //
+    //   - an eta-expanded function:
+    //     @implicitAmbiguous("msg A=${A}")
+    //     implicit def f[A](x: Int): String = ...
+    //     implicitly[Int => String] // found: x => f[Any](x)
+
+    val call = tpd.closureBody(alt.tree) // the tree itself if not a closure
+    val (_, targs, _) = tpd.decomposeCall(call)
+    val args = resolveTypes(targs)(using ctx.fresh.setTyperState(alt.tstate))
+    userDefinedErrorString(raw, params, args)
+  }
+
+  /** @param rawMsg           Message template with variables, e.g. "Variable A is ${A}"
+   *  @param sym              Symbol of the annotated type or of the method whose parameter was annotated
+   *  @param substituteType   Function substituting specific types for abstract types associated with variables, e.g A -> Int
+   */
+  private def formatAnnotationMessage(rawMsg: String, sym: Symbol, substituteType: Type => Type): String = {
+    val substitutableTypesSymbols = ErrorReporting.substitutableTypeSymbolsInScope(sym)
+
+    userDefinedErrorString(
+      rawMsg,
+      paramNames = substitutableTypesSymbols.map(_.name.unexpandedName.toString),
+      args = substitutableTypesSymbols.map(_.typeRef).map(substituteType)
+    )
+  }
+
+  /** Extracting the message from a method parameter, e.g. in
+   *
+   *  trait Foo
+   *
+   *  def foo(implicit @annotation.implicitNotFound("Foo is missing") foo: Foo): Any = ???
+   */
+  private def userDefinedImplicitNotFoundParamMessage = paramSymWithMethodCallTree.flatMap { (sym, applTree) =>
+    userDefinedMsg(sym, defn.ImplicitNotFoundAnnot).map { rawMsg =>
+      val methodOwner = applTree.symbol.owner
+      val methodOwnerType: Type = applTree match {
+        case Select(qual, _) => qual.tpe
+        case TypeApply(Select(qual, _), _) => qual.tpe
+        case _ => methodOwner.typeRef
+      }
```

Comment:
You can use `tpd.decomposeCall` to extract the method that is applied (as well as the type arguments which are used below), and then `tpd.qualifier` to get the prefix.